### DISCUSSION TITLE
CORS Issue, requests from angular app are blocked

### DISCUSSION THREAD

**Discussion by VukMNE** [strapi/strapi#7558] [Legacy Topics]
**Issue**
Hello,
I am getting my requests blocked by CORS when try to send requests from angular app.

**Error**

> Access to XMLHttpRequest at 'http://localhost:1337/posts?position.id=2&_start=0&_limit=3&_sort=updated_at:DESC' from origin 'http://localhost:4200' has been blocked by CORS policy: Request header field access-control-allow-origin is not allowed by Access-Control-Allow-Headers in preflight response.

**Steps to reproduce the behavior**

1. Start Strapi development mode
2. Start your Angular UI on same server
3. API calls to Strapi via Angular are blocked in browser by CORS

**Expected behavior**
After enabling CORS in Strapi and adding headers on Angular and or nginx the request should not be blocked via CORS.

**Code snippets**
Here are the examples of imporant files.

```
{
    "cors": {
    "enabled": true,
    "origin": "*",
    "expose": [
        "WWW-Authenticate",
        "Server-Authorization",
        "Access-Control-Expose-Headers"
     ],
    "maxAge": 31536000,
    "credentials": true,
    "methods": [
      "GET",
      "POST",
      "PUT",
      "PATCH",
      "DELETE",
      "OPTIONS",
      "HEAD"
    ],
    "headers": [
      "Content-Type",
      "Authorization",
      "X-Frame-Options",	  
      "Origin",
      "Access-Control-Allow-Headers",
      "access-control-allow-origin"
    ]
  }
}
```

**System**

- Node.js version: v10.16.3
- NPM version: v6.9.0
- Strapi version: 3.1.3
- Operating system: Windows 10

**Comment by derrickmehaffy** [comment]
The default settings for CORs shouldn't cause you issues, but it also looks like you are setting them in the wrong file.

**Reply by VukMNE** [reply]
@derrickmehaffy  Thank you for attention and responding. 

The file **security.json** that I wrote this setting is located in: config/environments/development

**Comment by derrickmehaffy** [comment]
Hello,

I see you are wanting to ask a question that is not really a bug report or a feature request, questions should be directed to the [discussions](https://github.com/strapi/strapi/discussions). Please see the following contributing guidelines for asking a question [here](https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md#reporting-an-issue).

I have converted this to a discussion for you, we try to keep Github clean and only for bug reports and feature requests.

**Comment by VukMNE** [comment]
@derrickmehaffy I undestand your disbelief, but I did everything according to the [strapi documentation](https://strapi.io/documentation/3.0.0-beta.x/concepts/configurations.html#security), and the requests from angular app are still getting blocked.

**Reply by derrickmehaffy** [reply]
I see the confusion here, you are using the old beta docs but are using the stable release, in which case the relevant docs are: https://strapi.io/documentation/v3.x/concepts/middlewares.html#security-middlewares

**Reply by VukMNE** [reply]
Thanks for the reply. 

Unfortunately, I still have the same problem.

**Here is my ./config/middleware.js**

```
module.exports = {
  settings: {
    cors: {
	  enabled: true,
      origin: ['*']
    }
  }
};
```

![error_strapi](https://user-images.githubusercontent.com/22847213/91147509-a7b13180-e6b8-11ea-9897-336b04faa98e.png)


Here is my strapi [project.](https://github.com/VukMNE/ztk-sajt-backend)

Here is my angular [app.](https://github.com/VukMNE/ztk-sajt)

If you run both of these projects on your computer, you will get the error from the picture above.

**Reply by derrickmehaffy** [reply]
Hi @VukMNE So in looking through both projects, in your `ztk-sajt/src/app/shared/services/data.service.ts` You have two headers set that shouldn't be sent from the angular app.

On lines 19 and 20:

```
  constructor(private http: HttpClient) {
    this.config = new Config();
    this.baseUrl = this.config.baseUrl;
    this.header = new HttpHeaders({
      'Content-Type': 'application/json',
      // 'Access-Control-Allow-Headers': '*',
      // 'Access-Control-Allow-Origin': '*'
    });
  }
```

You are sending two `Access-Control-Allow-*` headers from the angular app to Strapi, these should never be sent in a Request and are only send in a response. Angular should only send `Access-Control-Request-*` and Strapi will respond with `Access-Control-Allow-*` headers.

**Reply by VukMNE** [reply]
I see, thanks. It works smoothly without those headers now.

**Comment by petersg83** [comment]
Hi :)
I think you missed some parts in the migration guide: https://strapi.io/documentation/v3.x/migration-guide/migration-guide-beta.20-to-3.0.0.html#middlewares

Can you please move your `security.json` file to `./config/env/development/middleware.js` ?

Update content:
```
module.exports = {
  settings: {
    cors: {
      enabled: true,
      origin: '*',
      expose: ['WWW-Authenticate', 'Server-Authorization', 'Access-Control-Expose-Headers'],
      maxAge: 31536000,
      credentials: true,
      methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS', 'HEAD'],
      headers: [
        'Content-Type',
        'Authorization',
        'X-Frame-Options',
        'Origin',
        'Access-Control-Allow-Headers',
        'access-control-allow-origin',
      ],
    },
  },
};
```

**Reply by VukMNE** [reply]
Thanks. After moving file to this location, it is working smoothly now :)

**Reply by ducdk** [reply]
i config in `./config/env/product/middleware.js` but not working

**Comment by derrickmehaffy** [comment]
Locking as migrated to: https://forum.strapi.io/t/cors-issue-requests-from-angular-app-are-blocked-7558/1537
